File: backend/src/models/Workflow.js
import mongoose from 'mongoose'

import createSchema from './utils/createSchema'
import generateId from '../shared/utils/generateId'
import {ACCESS_ROLES, MEDIA_POSITIONS} from '../shared/config/constants'

const {Types} = mongoose.Schema

export const NodeTag = createSchema(
  {
    id: String,
    name: String,
    color: String,
  },
  {_id: false, timestamps: false},
)

export const Node = createSchema(
  {
    id: {type: String, required: true},
    children: {type: [String]},
    image: Types.ObjectId,
    imagePosition: {type: String, enum: Object.values(MEDIA_POSITIONS)},
    video: {type: String, required: false},
    file: Types.ObjectId,
    title: String,
    collapsed: Boolean,
    color: String,
    borderColor: String,
    scale: Number,
    parent: String,
    x: Number,
    y: Number,
    width: Number,
    height: Number,
    tags: {type: [String]},
    autoshrink: {type: Boolean, default: false},
    command: {type: String, required: false},
    prompts: {type: [String]},
  },
  {_id: false, timestamps: false},
)

export const Edge = createSchema(
  {
    id: {type: String, required: true},
    start: String,
    end: String,
    title: String,
    color: String,
  },
  {_id: false, timestamps: false},
)

const RoleBinding = createSchema(
  {
    subjectId: {type: String, required: true},
    subjectType: {type: String, enum: ['user', 'group', 'mail'], required: true},
    role: {type: String, enum: Object.values(ACCESS_ROLES), required: true},
  },
  {_id: false, timestamps: false},
)

const Share = createSchema(
  {
    public: {
      enabled: Boolean,
      hidden: Boolean,
      writeable: Boolean,
    },
    // link: {
    //   id: String,
    //   password: {type: String, select: false},
    //   until: {type: Types.Datetime},
    // },
    access: [RoleBinding],
  },
  {_id: false, timestamps: false},
)

// const config = {
//   titleLeftMargin: 13,
//   titleRightMargin: 5,
//   titleTopMargin: 15,
//   titleBottomMargin: 5,
//   titleJustifyLines: 4,
//   roundCornerWidth: 5,
//   headerLineMargin: 13,
//   oddColor: '#ffffff',
//   oddOpacity: 0.44,
//   zoomMargin: 20,
//   edgeTextDistance: 2,
// }
//     config: {
//       titleLeftMargin: Number,
//       titleRightMargin: Number,
//       titleTopMargin: Number,
//       titleBottomMargin: Number,
//       titleJustifyLines: Number,
//       roundCornerWidth: Number,
//       headerLineMargin: Number,
//       oddColor: String,
//       oddOpacity: Number,
//       zoomMargin: Number,
//       edgeTextDistance: Number,
//     },

const WorkflowSchema = createSchema({
  mapId: {type: String, index: true, default: generateId},
  userId: {type: String, required: true, index: true},
  title: String,
  nodes: {type: Map, of: Node},
  edges: {type: Map, of: Edge},
  root: String,
  tags: {type: [NodeTag]},
  share: Share,
  category: String,
})

WorkflowSchema.methods.isPublic = function () {
  return Boolean(this.share?.public?.enabled)
}
WorkflowSchema.methods.isPublicWriteable = function () {
  return Boolean(this.share?.public?.enabled && this.share?.public?.writeable)
}

const Workflow = mongoose.model('Workflow', WorkflowSchema)

export default Workflow

File: backend-v2/internal/models/workflow.go
package models

import (
	"backend-v2/internal/common/constants"
)

type MediaPosition string

const (
	MediaCrop      MediaPosition = "crop"
	MediaBody      MediaPosition = "body"
	MediaStretch   MediaPosition = "stretch"
	MediaFullWidth MediaPosition = "fullWidth"
)

type Node struct {
	ID            string        `json:"id" bson:"id"`
	Children      []string      `json:"children" bson:"children"`
	Image         string        `json:"image" bson:"image"`
	ImagePosition MediaPosition `json:"imagePosition" bson:"imagePosition"`
	Video         string        `json:"video" bson:"video"`
	File          string        `json:"file" bson:"file"`
	Title         string        `json:"title" bson:"title"`
	Collapsed     bool          `json:"collapsed" bson:"collapsed"`
	Color         string        `json:"color" bson:"color"`
	BorderColor   string        `json:"borderColor" bson:"borderColor"`
	Scale         int           `json:"scale" bson:"scale"`
	Parent        string        `json:"parent" bson:"parent"`
	X             int64         `json:"x" bson:"x"`
	Y             int64         `json:"y" bson:"y"`
	Width         int           `json:"width" bson:"width"`
	Height        int           `json:"height" bson:"height"`
	Command       string        `json:"command" bson:"command"`
	Prompts       []string      `json:"prompts" bson:"prompts"`
}

type Edge struct {
	ID    string `json:"id" bson:"id"`
	Start string `json:"start" bson:"start"`
	End   string `json:"end" bson:"end"`
	Title string `json:"title" bson:"title"`
	Color string `json:"color" bson:"color"`
}

type RoleBinding struct {
	SubjectID   string                `json:"subjectId" bson:"subjectId"`
	SubjectType constants.SubjectType `json:"subjectType" bson:"subjectType"`
	Role        constants.AccessRole  `json:"role" bson:"role"`
}

type WorkflowState struct {
	Enabled   bool `json:"enabled" bson:"enabled"`
	Hidden    bool `json:"hidden" bson:"hidden"`
	Writeable bool `json:"writeable" bson:"writeable"`
}

type Share struct {
	Public WorkflowState `json:"public" bson:"public"`
	Access []RoleBinding `json:"access" bson:"access"`
}

type Workflow struct {
	UserID     string            `json:"userId" bson:"userId"`
	WorkflowID string            `json:"workflowId" bson:"workflowId"`
	Title      string            `json:"title" bson:"title"`
	UpdatedAt  int64             `json:"updatedAt" bson:"updatedAt"`
	Nodes      map[string]Node   `json:"nodes" bson:"nodes"`
	Edges      map[string]Edge   `json:"edges" bson:"edges"`
	Files      map[string]string `json:"files" bson:"files"`
	Share      Share             `json:"share" bson:"share"`
	Category   *string           `json:"category" bson:"category"`
}

func (w Workflow) IsPublic () bool {
	return  w.Share.Public.Enabled
}

func (w Workflow) IsPublicWriteable () bool {
	return  w.Share.Public.Enabled && w.Share.Public.Writeable
}